program/src/processor/deactivate_stake.rs

```diff
@@ -1,21 +1,135 @@
+#[inline(always)]
+fn get_max_deactivation_amount(total_amount: u64, basis_points: u16) -> Result<u64, ProgramError> {
+    total_amount
+        .checked_mul(basis_points as u64)
+        .and_then(|p| p.checked_div(10_000))
+        .ok_or(ProgramError::ArithmeticOverflow)
+}
 
 /// Deactivate staked tokens for the validator.
 ///
 /// Only one deactivation may be in-flight at once, so if this is called
 /// with an active deactivation, it will succeed, but reset the amount and
 /// timestamp.
 ///
-/// 0. `[w]` Validator stake account
-/// 1. `[s]` Authority on validator stake account
+/// 0. `[]` Stake config account
+/// 1. `[w]` Validator stake account
+/// 2. `[s]` Authority on validator stake account
 ///
 /// Instruction data: amount of tokens to deactivate, as a little-endian u64
 pub fn process_deactivate_stake(
-    _program_id: &Pubkey,
-    _ctx: Context<DeactivateStakeAccounts>,
-    _amount: u64,
+    program_id: &Pubkey,
+    ctx: Context<DeactivateStakeAccounts>,
+    amount: u64,
 ) -> ProgramResult {
+    // Account validation.
+
+    // config
+    // - owner must be the stake program
+    // - must be initialized
+
+    require!(
+        ctx.accounts.config.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "config"
+    );
+
+    let data = &ctx.accounts.config.try_borrow_data()?;
+    let config = bytemuck::try_from_bytes::<Config>(data)
+        .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    require!(
+        config.is_initialized(),
+        ProgramError::UninitializedAccount,
+        "config",
+    );
+
+    // stake
+    // - owner must be the stake program
+    // - must be initialized
+    // - must have the correct derivation
+
+    require!(
+        ctx.accounts.stake.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "stake"
+    );
+
+    let data = &mut ctx.accounts.stake.try_borrow_mut_data()?;
+    let stake = bytemuck::try_from_bytes_mut::<Stake>(data)
+        .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    require!(
+        stake.is_initialized(),
+        ProgramError::UninitializedAccount,
+        "stake",
+    );
+
+    // validates that the stake account corresponds to the received config account
+    let (derivation, _) = find_stake_pda(&stake.validator, ctx.accounts.config.key, program_id);
+
+    require!(
+        ctx.accounts.stake.key == &derivation,
+        ProgramError::InvalidSeeds,
+        "stake",
+    );
+
+    // authority
+    // - must be a signer
+    // - must match the authority on the stake account
+
+    require!(
+        ctx.accounts.stake_authority.is_signer,
+        ProgramError::MissingRequiredSignature,
+        "stake_authority"
+    );
+
+    require!(
+        ctx.accounts.stake_authority.key == &stake.authority,
+        StakeError::InvalidAuthority,
+        "stake_authority"
+    );
+
+    // Validate the amount.
+
+    require!(amount > 0, StakeError::InvalidAmount);
```

Comment:
I'm wondering if it's worth allowing 0 -- that means someone can cancel a deactivation. What do you think?

Reply:
Hmmm, that would look like a side-effect so not sure whether we should allow that or create an specific instruction for that. Also, there is no "effect" of deactivating unless you move the tokens to inactive.

Reply:
Moving the tokens to inactive is a permissionless instruction, so if I accidentally deactivate, then I need to be ready to re-activate right when the time passes if I don't want to lose out on any rewards. With the current behavior, I would probably override my incorrect deactivation with another deactivation for 1 token, which is kind of silly.

If you want to add a new instruction, that's fine by me! But that would be a many-line change vs a one-line change.

Reply:
Good point, I will change so an amount of `0` cancels the deactivation.